### ISSUE TITLE
Ginormous Winetricks gui

### ISSUE THREAD

**Issue by UltraBlackLinux** [sonic2kk/steamtinkerlaunch#319]
Hey there,
I've noticed a problem with the winetricks GUI:
![video](https://user-images.githubusercontent.com/62404294/136691607-704b6a0b-c312-43a6-aa9a-5b71e424b93a.mp4)
Yeah... Please fix this xD
Thanks!

**Comment by frostworx** [comment]
You do not even test/react on features which were implemented for you, that's why you're on my spam/ignore list
https://github.com/frostworx/steamtinkerlaunch/issues/298
https://github.com/frostworx/steamtinkerlaunch/issues/283

Expecting much and not giving back anything.

**Comment by UltraBlackLinux** [comment]
> You do not even test/react on features which were implemented for you, that's why you're on my spam/ignore list #298 #283
> 
> Expecting much and not giving back anything.

Bruh sorry. I don't recieve notifications for closed issues. The things I have requested have worked perfectly - I just forgot to give you some status updates (because of notifs again.)
I have already told you about this multiple times - I am not trying to be an asshole here. 
![Screenshot_211](https://user-images.githubusercontent.com/62404294/136692042-2916898c-9595-413d-9597-6918c024a251.png)

**Comment by frostworx** [comment]
Ok, accepted.
I was really annoyed that I did not even get a reply when spending my time on feature requests...

The giant gui is a yad bug, and no stl problem _(usually you should be able to at least scale down the window to the allowed minimum to make it look a bit better, which doesn't fix the core problem of course)_
I [already reported the bug upstream](https://github.com/v1cont/yad/issues/136) a while ago, but looks like upstream is busy with RL _([so am I very soon](https://github.com/frostworx/steamtinkerlaunch/issues/309))_
There are several other bug reports upstream which describe the same bug. It might help to test yad current master, as it is [said to be fixed there](https://github.com/v1cont/yad/issues/147#issuecomment-914213781).

**Comment by frostworx** [comment]
In this case the high boxes likely are a result from the pretty long winetricks packages list btw.

**Comment by UltraBlackLinux** [comment]
> Ok, accepted. I was really annoyed that I did not even get a reply when spending my time on feature requests...
> 
> The giant gui is a yad bug, and no stl problem _(usually you should be able to at least scale down the window to the allowed minimum to make it look a bit better, which doesn't fix the core problem of course)_ I [already reported the bug upstream](https://github.com/v1cont/yad/issues/136) a while ago, but looks like upstream is busy with RL _([so am I very soon](https://github.com/frostworx/steamtinkerlaunch/issues/309))_ There are several other bug reports upstream which describe the same bug. It might help to test yad current master, as it is [said to be fixed there](https://github.com/v1cont/yad/issues/147#issuecomment-914213781).

Thank you.
Could you maybe also improve the Winetricks wiki page? I don't quite know, how to use the winetricks gui, like the select and install button (The names don't show anymore on hovering over top of buttons and checkboxes so I don't know what setting does what)

**Comment by frostworx** [comment]
Good idea. I guess the upper part of the gui is self-explanatory, but only the main SELECT, DONE and INSTALL buttons are not, right?
With above "In this case the high boxes likely are a result from the pretty long winetricks packages list btw." I meant
the "Currently selected packages" list at the bottom btw, they trigger the box bug.

**Comment by UltraBlackLinux** [comment]
> but only the main SELECT, DONE and INSTALL buttons are not, right?

Yeah, I didn't quite understand those

> With above "In this case the high boxes likely are a result from the pretty long winetricks packages list btw." I meant the "Currently selected packages" list at the bottom btw, they trigger the box bug.

I got that ;)

**Comment by frostworx** [comment]
Updated the wiki, I hope it is better understandable now:
https://github.com/frostworx/steamtinkerlaunch/wiki/Winetricks

**Comment by UltraBlackLinux** [comment]
@frostworx Thank you! It's way clearer now!
(Btw, I found a typo: "Using the INSTALL buton" buton) 
One last thing though: Does the "Save winetricks proton" button install all selected packages into the selected proton version permanently?

**Comment by frostworx** [comment]
@typo. fixed, thx.
By default the proton/wine version used for installing packages into the wineprefix is the same which the game currently uses.
The gui offers the option to override the "winetricks" version via dropdown menu.
If the optional "Save winetricks proton" box is checked, the selected override will be used automatically by default in this gui to INSTALL additional packages later _(so the user doesn't have to remember which version was used before and doesn't have to set it again)_

_(will add this answer to the wiki as well)_ Edit: *done*

**Comment by UltraBlackLinux** [comment]
> @typo. fixed, thx. By default the proton/wine version used for installing packages into the wineprefix is the same which the game currently uses. The gui offers the option to override the "winetricks" version via dropdown menu. If the optional "Save winetricks proton" box is checked, the selected override will be used automatically by default in this gui to INSTALL additional packages altlaterer _(so the user doesn't have to remember which version was used before and doesn't have to set it again)_
> 
> _(will add this answer to the wiki as well)_

Ah okay, that makes sense.
Well there are just more and more things that I want to address xD
... The "Save winetricks proton" button duplicates the proton version in the proton-list - Is there a way to revert that/delete a specific proton version?

**Comment by frostworx** [comment]
> ... The "Save winetricks proton" button duplicates the proton version in the proton-list - Is there a way to revert that/delete a specific proton version?

It should be possible, but this would have to be changed in multiple dropdown menus.
Should be even possible with a little generic function, but I'm afraid I don't have the time anymore to change that.
TL;DR: maybe :)

**Comment by UltraBlackLinux** [comment]
> > ... The "Save winetricks proton" button duplicates the proton version in the proton-list - Is there a way to revert that/delete a specific proton version?
> 
> It should be possible, but this would have to be changed in multiple dropdown menus. Should be even possible with a little generic function, but I'm afraid I don't have the time anymore to change that. TL;DR: maybe :)

Ah okay.
Thanks for your assistance!

**Comment by frostworx** [comment]
A quick and dirty
`		--field="$GUI_WTPROTON!$(strFix "$DESC_WTPROTON" "$BUT_INSTALL")":CB "${WTPROTON//\"}!${PROTYADLIST//!${WTPROTON//\"}}" \`

[here](https://github.com/frostworx/steamtinkerlaunch/blob/master/stl#L9646)

would fix this in most cases, but not when the first available option is the current selection

**Comment by UltraBlackLinux** [comment]
> A quick and dirty ` --field="$GUI_WTPROTON!$(strFix "$DESC_WTPROTON" "$BUT_INSTALL")":CB "${WTPROTON//\"}!${PROTYADLIST//!${WTPROTON//\"}}" \`
> 
> [here](https://github.com/frostworx/steamtinkerlaunch/blob/master/stl#L9646)
> 
> would fix this in most cases, but not when the first available option is the current selection

That's chinese for me now at this point xD

**Comment by frostworx** [comment]
Hehe.
Looks like it fixing this with a generic function is pretty simple.
Just added it for winetricks. Please test.
Will add it for all other dropdowns as well then.

**Comment by frostworx** [comment]
oh, wait a sec, need to push the commit first :)

**Comment by frostworx** [comment]
https://github.com/frostworx/steamtinkerlaunch/commit/78e31e4530a15fec8c36455d612c296b77f8acc8 should fix it for winetricks

**Comment by UltraBlackLinux** [comment]
> [78e31e4](https://github.com/frostworx/steamtinkerlaunch/commit/78e31e4530a15fec8c36455d612c296b77f8acc8) should fix it for winetricks
![Screenshot_212](https://user-images.githubusercontent.com/62404294/136694639-5750544b-efad-4b70-be26-b70e9891b2e0.png)
It kinda fixed it, but only for a few entries, this is what it looks like in the proton override option: ![Screenshot_213](https://user-images.githubusercontent.com/62404294/136694698-a150a7db-047f-4cfb-af45-d549b01ec973.png)

**Comment by frostworx** [comment]
thanks for testing. Yes, the current commit only fixes the winetricks gui, just fixing all other _(~55)_ dropdowns

**Comment by UltraBlackLinux** [comment]
> thanks for testing. Yes, the current commit only fixes the winetricks gui, just fixing all other _(~55)_ dropdowns

Well I rather meant that as a comparison between before-and-after. As the screenshot shows, the first two entries got fixed, but all of the others seem to be the same as before

**Comment by frostworx** [comment]
Hah, that is a different problem. I thought you mean that the current selection is listed twice in the dropdown _(which I'm about ti fix just right now)_
For your problem:
The list of available proton versions is generated once initially _(and hopefully always when a proton version was added)_.
The temp list can be found under `/dev/shm/stl/ProtonCSV.txt`
Chances are high, that you have the same Proton version in multiple places and therefore it is listed twice.
Of course removing duplicates would be possible here as well, but the user will never see that duplicates are even installed, that's why I leave this as is.

**Comment by frostworx** [comment]
"removed the duplicate of the current selection in all dropdown menus" added in https://github.com/frostworx/steamtinkerlaunch/commit/a5e8b5fca21861c1fc20b2c7b612d7eee996dffc

**Comment by UltraBlackLinux** [comment]
> Hah, that is a different problem. I thought you mean that the current selection is listed twice in the dropdown _(which I'm about ti fix just right now)_ For your problem: The list of available proton versions is generated once initially _(and hopefully always when a proton version was added)_. The temp list can be found under `/dev/shm/stl/ProtonCSV.txt` Chances are high, that you have the same Proton version in multiple places and therefore it is listed twice. Of course removing duplicates would be possible here as well, but the user will never see that duplicates are even installed, that's why I leave this as is.

Oh yeah, I see. This is very annoying, as steam has a local root folder pointing to the steam folder (In my case it's on a separate drive) - That's another compatibility thing probably. - and stl detects the config folder, but also the drive itself:
```
proton-3.16-9;/home/maus/.steam/root/steamapps/common/Proton 3.16 Beta/proton
proton-3.16-9;/run/media/maus/Steam/steamapps/common/Proton 3.16 Beta/proton
```
Could you try to exclude /run/ devices when creating this file? The steam backwards compatibility hacks are probably never going to die

**Comment by frostworx** [comment]
This doesn't look like default behavior. Are any symlinks involved here?
If so please carefully list the exact symlinks involved, so I might be able _(can't promise anything)_
to remove any duplicates in the proton list, caused by symlinks.

**Comment by UltraBlackLinux** [comment]
> This doesn't look like default behavior. Are any symlinks involved here? If so please carefully list the exact symlinks involved, so I might be able _(can't promise anything)_ to remove any duplicates in the proton list, caused by symlinks.

```
╭─maus@lolcat in ~/.steam
╰─λ ls -la .
lrwxrwxrwx   23 maus 10 Oct 12:12  bin -> /home/maus/.steam/bin32
lrwxrwxrwx   33 maus 10 Oct 12:12  bin32 -> /run/media/maus/Steam/ubuntu12_32
lrwxrwxrwx   33 maus 10 Oct 12:12  bin64 -> /run/media/maus/Steam/ubuntu12_64
lrwxrwxrwx   21 maus 10 Oct 12:12  root -> /run/media/maus/Steam
lrwxrwxrwx   29 maus 10 Oct 12:12  sdk32 -> /run/media/maus/Steam/linux32
lrwxrwxrwx   29 maus 10 Oct 12:12  sdk64 -> /run/media/maus/Steam/linux64
lrwxrwxrwx   21 maus 10 Oct 12:12  steam -> /run/media/maus/Steam
.rwxr-xr-x 6.0k maus 10 Oct 14:04  registry.vdf
.rwxr-xr-x  204 maus 19 Jun 12:43  registry.vdf.bak
.rwxr-xr-x 2.9k maus 12 May 22:39  registry.vdf20200921204158
.rwxr-xr-x 2.9k maus 12 May 22:39  registry.vdf20200926222113
.rwxr-xr-x 3.3k maus 12 May 22:39  registry.vdf20201031212736
.rwxr-xr-x 3.5k maus 12 May 22:39  registry.vdf20201120232627
.rwxr-xr-x 3.8k maus 12 May 22:39  registry.vdf20201221162913
.rwxr-xr-x 4.0k maus 12 May 22:39  registry.vdf20210405213854
.rwxr-xr-x 4.2k maus 12 May 22:39  registry.vdf20210413215033
.rwxr-xr-x 4.6k maus 26 May 22:21  registry.vdf20210526222117
.rwxr-xr-x 4.2k maus  1 Aug 12:39  registry.vdf20210801123923
.rw-r--r--    7 maus 10 Oct 12:12  steam.pid
|rw-------    0 maus  3 Oct 16:22  steam.pipe
.r--------   16 maus 10 Oct 12:12  steam.token
```
I can't think of anywhere else... - Where is STL searching?

**Comment by frostworx** [comment]
Weird, I assume you didn't create those symlinks manually, did you?
Which distro is this?
Should be fixable, but not today. RL calls.

**Comment by UltraBlackLinux** [comment]
> Weird, I assume you didn't create those symlinks manually, did you? Which distro is this? Should be fixable, but not today. RL calls.

I did not create these symlinks. I might just rename that folder and have steam recreate that. It's old as heck. I think I switched my steam folder 1-2 years ago on Manjaro. A few months ago I switched to Garuda Linux

EDIT: I really can't figure it out :/

**Comment by frostworx** [comment]
udisks2 automounts to `/run/media`, mayve this is not even the problem.
You wrote that the steam local root folder is "on a separate drive" in your case. Means you changed the path manually, right?
Do you have that separate drive also as additional steam library path? If So this would explain the duplicates.

**Comment by UltraBlackLinux** [comment]
> udisks2 automounts to `/run/media`, mayve this is not even the problem. You wrote that the steam local root folder is "on a separate drive" in your case. Means you changed the path manually, right? Do you have that separate drive also as additional steam library path? If So this would explain the duplicates.

I didn't do it manually manually, I changed it in steam, back then when there was the old gui - I don't know how to do it in the new one, since I removed the /home/ entry last time, but in the new gui you can't do that anymore (I have no clue)?
![Screenshot_221](https://user-images.githubusercontent.com/62404294/136705840-f99da7f1-eac9-4ff3-9875-69b7cbf1d313.png)

**Comment by frostworx** [comment]
No problem.
`readlink -f ~/.steam/steam/compatibilitytools.d` does point to the corresponding /run/media path on your system, right?

**Comment by UltraBlackLinux** [comment]
> No problem. `readlink -f ~/.steam/steam/compatibilitytools.d` does point to the corresponding /run/media path on your system, right?

exactly.

**Comment by frostworx** [comment]
ok, then I'll simply rewrite all related parts of `getAvailableProtonVersions` to always use the "readlink -f path", and filter duplicates afterwards _(maybe not today)_

**Comment by frostworx** [comment]
This commit https://github.com/frostworx/steamtinkerlaunch/commit/3bb44a79fdb7c3aa90b17aad46458d199028ac4a should fix it already.

~~The previous fix for removing duplicates in dropdown menus leads to a _(harmless?)_
`GLib-WARNING g_strcompress: trailing \` `yad` warning now, which requires a fix as well.~~ *fixed*

**Comment by UltraBlackLinux** [comment]
> This commit [3bb44a7](https://github.com/frostworx/steamtinkerlaunch/commit/3bb44a79fdb7c3aa90b17aad46458d199028ac4a) should fix it already.
> 
> ~The previous fix for removing duplicates in dropdown menus leads to a _(harmless?)_ `GLib-WARNING g_strcompress: trailing \` `yad` warning now, which requires a fix as well.~ _fixed_

Oh, seems like your attempt at fixing this wasn't too successful :/
```
experimental-6.3-20210930;/home/maus/.steam/root/steamapps/common/Proton - Experimental/proton
proton-6.3-7;/home/maus/.steam/root/steamapps/common/Proton 6.3/proton
proton-3.16-9;/home/maus/.steam/root/steamapps/common/Proton 3.16 Beta/proton
proton-5.13-6b;/home/maus/.steam/root/steamapps/common/Proton 5.13/proton
experimental-6.3-20210930;/run/media/maus/Steam/steamapps/common/Proton - Experimental/proton
proton-6.3-7;/run/media/maus/Steam/steamapps/common/Proton 6.3/proton
proton-3.16-9;/run/media/maus/Steam/steamapps/common/Proton 3.16 Beta/proton
proton-5.13-6b;/run/media/maus/Steam/steamapps/common/Proton 5.13/proton
Proton-6.15-GE-2;/home/maus/.config/stl/proton/custom/Proton-6.15-GE-2/proton
Proton-6.18-GE-2;/home/maus/.config/stl/proton/custom/Proton-6.18-GE-2/proton
```
The problem stil remains - Apparently your check didn't trigger:

```
Mon 11 Oct 08:04:53 CEST 2021 INFO - listAllSettingsEntries - '/dev/shm/stl/setentries.txt' already exists - nothing to do
Mon 11 Oct 08:04:53 CEST 2021 INFO - createDefaultCfgs - STOP
Mon 11 Oct 08:04:54 CEST 2021 INFO - initOldProtonArr - Creating an initial array with available Proton versions using the file '/dev/shm/stl/ProtonCSV.txt' which was created during a previous run
Mon 11 Oct 08:04:54 CEST 2021 INFO - initFirstProton - Initializing Proton
Mon 11 Oct 08:04:54 CEST 2021 INFO - initFirstProton - Initial Proton version 'FIRSTUSEPROTON' from '/home/maus/.config/stl/default_template.conf' is 'proton-6.3-7'
Mon 11 Oct 08:04:54 CEST 2021 INFO - getGameName - Found Game Name '"BeamNG.drive"' in '/home/maus/.config/stl/gamecfgs/id/284160.conf'
```

Oh I see. You are checking if the file contains the whole line (Which is never true when there are two locations), and not if the name already exists: This fixes it, I don't know how to open pull requests xD
`1731		if [[ ! " ${ProtonCSV[*]} " =~ $protonfileV ]]; then`

**Comment by UltraBlackLinux** [comment]
> In this case the high boxes likely are a result from the pretty long winetricks packages list btw.

Regarding this - Four packages and still fat buttons.
The size of them seems to actually scale with the amount of text at the bottom - Compare this with the video: 
![Screenshot_223](https://user-images.githubusercontent.com/62404294/136745397-0612ad5d-e8d2-4a1b-89cf-daf8e60a7020.png)

**Comment by frostworx** [comment]
> Oh I see. You are checking if the file contains the whole line (Which is never true when there are two locations), and not if the name already exists: 

Nah, that's why I compare the` readlink paths` and not the original paths - this should be fine.
Your duplicates probably just came from an old `/dev/shm/stl/ProtonCSV.txt`.

Anyway `find` doesn't follow symlinks by default which needs to be fixed - else some proton version might not be found, so this needs some more attention anyway.

> This fixes it, I don't know how to open pull requests xD

bad excuse!

> `1731 if [[ ! " ${ProtonCSV[*]} " =~ $protonfileV ]]; then`

should work as well, but comparing the whole line `$PCSV` should be fine anyway

**Comment by frostworx** [comment]
The core problem of your duplicates is simply that the 
_(recently added)_ `~/.steam/steam/config/libraryfolders.vdf` contains the base `~/.local/share/Steam` as well,
where the previously used `~/.steam/steam/config/config.vdf` doesn't.
Since someone reported, that their paths are only in `libraryfolders.vdf`, I check both everywhere.

_(`grep "BaseInstallFolder\|\"path\"" ~/.steam/steam/config/libraryfolders.vdf`
vs
`grep "BaseInstallFolder\|\"path\"" ~/.steam/steam/config/config.vdf`)_

A simple fix would be to remove the whole `DEFSTEAMAPPSCOMMON` check, but only if it is safe to assume that everybody has the base `~/.local/share/Steam` in their `config.vdf`. Until then, I'll add a check in `DEFSTEAMAPPSCOMMON`

A filter for possible duplicates caused by _(likely user created)_ symlinks is good to have anyway.

**Comment by UltraBlackLinux** [comment]
> > Oh I see. You are checking if the file contains the whole line (Which is never true when there are two locations), and not if the name already exists:
> 
> Nah, that's why I compare the` readlink paths` and not the original paths - this should be fine. Your duplicates probably just came from an old `/dev/shm/stl/ProtonCSV.txt`.

Nope. I deleted ProtonCSV.txt before trying my fix, and I can confirm, that it wasn't working before I put it in place.

**Comment by frostworx** [comment]
https://github.com/frostworx/steamtinkerlaunch/commit/96ba75101ea655fec7da362736f423840410f712 looks good

- use the original path of all autodetected steam paths generally _(might be helpful to be more distro independant)_
- follow symlinks to find proton installations _(and therefore trigger the previously duplicate removal everywhere)_
- skip the separate proton check in base `~/.local/share/Steam` if `config.vdf` has it as `path` as well

This and changes above should cover every possible combination of possible proton duplicates
_(of course it still can happen that the `version` in the proton shipped version files have duplicates, so duplicates in the short list (dropdown) can still be possible, but they have different paths)_

**Comment by UltraBlackLinux** [comment]
> [96ba751](https://github.com/frostworx/steamtinkerlaunch/commit/96ba75101ea655fec7da362736f423840410f712) looks good
> 
>     * use the original path of all autodetected steam paths generally _(might be helpful to be more distro independant)_
> 
>     * follow symlinks to find proton installations _(and therefore trigger the previously duplicate removal everywhere)_
> 
>     * skip the separate proton check in base `~/.local/share/Steam` if `config.vdf` has it as `path` as well
> 
> 
> This and changes above should cover every possible combination of possible proton duplicates _(of course it still can happen that the `version` in the proton shipped version files have duplicates, so duplicates in the short list (dropdown) can still be possible, but they have different paths)_

![Screenshot_224](https://user-images.githubusercontent.com/62404294/136760507-fac68873-200a-489a-b910-facb96ca27bf.png)
Works!
Close?

**Comment by frostworx** [comment]
.
